### PR TITLE
Fixes clipping

### PR DESCRIPTION
**Pull request by ananyahjha93**
Added tests (CPU and GPU) to compare torch clipping and olmo clipping and fixed clipping for DDP and FSDP no_shard

@epwalsh can we merge this PR so that I can push the DDP one after this?

### CODE REVIEW

File: olmo/optim.py [hunk 1]
```diff
@@ -195,27 +199,31 @@ def is_grad_norm_metric(metric_name: str) -> bool:
...
+        #######################################################################
+        # part 3: clip grads
+        #######################################################################
         num_grads_clipped = 0
         num_eligible_grads = 0
         for group in self.param_groups:
             if (max_norm_ratio := group.get("max_grad_norm_ratio")) is not None:
                 num_clipped = self._do_adaptive_clipping(
-                    group, max_norm_ratio, global_step, all_metrics, collect_param_metrics=collect_param_metrics
+                    group, max_norm_ratio, global_step, all_metrics, collect_param_metrics=True
```
Comment (epwalsh) [inline olmo/optim.py:222]: Why this change? This will force a host device sync on every batch.

File: olmo/optim.py [hunk 2]
```diff
@@ -195,27 +199,31 @@ def is_grad_norm_metric(metric_name: str) -> bool:
...
                 )
             elif (max_norm := group.get("max_grad_norm")) is not None:
                 num_clipped = self._do_global_fixed_clipping(
-                    group, max_norm, all_metrics, collect_param_metrics=collect_param_metrics
+                    group, max_norm, all_metrics, collect_param_metrics=True
```
Comment (epwalsh) [inline olmo/optim.py:226]: Same question here.

**Review by epwalsh — COMMENTED** [review]
@ananyahjha93 please don't merge anything into `main` without a review. This PR touches some critical components and I'm pretty sure your changes will force an unnecessary host device sync on every batch, which could potentially have a big negative impact on training throughput.

File: olmo/optim.py (hunk 1 above)
Comment (ananyahjha93) [reply olmo/optim.py:222]: let me send in a PR for this fix!
